File: python-sidecar/app/taxonomy.py
from __future__ import annotations

TAG_TAXONOMY: dict[str, list[str]] = {
    "People": ["person", "people", "portrait", "group", "selfie", "baby", "family",
               "man", "woman", "child", "kid", "toddler", "elderly", "couple", "crowd", "friends"],
    "Nature": ["nature", "landscape", "mountain", "forest", "tree", "flower", "plant", "garden",
               "desert", "jungle", "canyon", "field", "meadow", "hills", "cliff", "grass", "rock"],
    "Water": ["ocean", "beach", "sea", "lake", "river", "water", "underwater", "wave", "coral",
              "waterfall", "swimming-pool", "pond", "stream", "creek"],
    "Sky & Light": ["sunset", "sunrise", "sky", "cloud", "night", "moon", "star", "milky-way", "aurora",
                    "golden-hour", "blue-hour", "twilight", "morning", "afternoon", "evening",
                    "sunlight", "lens-flare", "bokeh", "backlight"],
    "Urban": ["city", "urban", "building", "architecture", "bridge", "street", "road", "alley",
              "skyline", "indoor", "room", "house", "home", "office", "warehouse", "studio",
              "subway", "skyscrapers", "monument", "cafe", "restaurant", "living-room", "bedroom"],
    "Food": ["food", "drink", "meal", "kitchen", "coffee", "dessert", "fruit",
             "bread", "vegetables", "meat", "rice", "pasta", "pizza", "burger", "sushi", "tea", "beer", "wine"],
    "Animals": ["animal", "pet", "dog", "cat", "bird", "horse", "wildlife", "fish", "butterfly",
                "insect", "reptile", "deer", "squirrel", "lion", "tiger", "bear", "elephant"],
    "Vehicles": ["vehicle", "car", "bicycle", "airplane", "train", "boat", "motorcycle", "truck",
                 "bus", "scooter", "ship", "yacht", "helicopter"],
    "Sport": ["sport", "fitness", "running", "biking", "soccer", "basketball", "swimming", "yoga",
              "fitness-model", "bodybuilder", "muscle", "athletic",
              "tennis", "golf", "badminton", "baseball", "gym", "climbing", "surfing", "skiing", "skateboarding"],
    "Art": ["art", "museum", "statue", "painting", "sculpture", "graffiti", "tattoo",
            "drawing", "sketch", "illustration", "digital-art", "crafts", "calligraphy"],
    "Events": ["event", "concert", "wedding", "party", "celebration", "festival", "parade",
               "concert-stage", "birthday", "anniversary", "graduation", "performance"],
    "Travel": ["travel", "vacation", "adventure", "camping", "road-trip", "hiking",
               "luggage", "passport", "sightseeing", "hotel", "resort", "map", "backpack"],
    "Technology": ["technology", "computer", "phone", "screen", "robot", "drone", "gadget",
                   "smartphone", "laptop", "keyboard", "mouse", "headphones", "camera", "tv"],
    "Abstract": ["abstract", "pattern", "texture", "colorful", "dark", "light", "geometric", "symmetry",
                 "minimalist", "vintage", "retro", "modern", "classic", "industrial", "rustic", "grunge",
                 "gradient", "smoke", "fire", "water-drops", "neon"],
    "Weather": ["rain", "snow", "fog", "storm", "rainbow", "lightning", "mist",
                "sunny", "windy", "snowy", "rainy", "foggy", "cloudy"],
    "Photography": ["macro", "aerial", "panorama", "reflection", "silhouette", "shadow", "long-exposure",
                    "macro-lens", "wide-angle", "telephoto", "drone-shot", "fish-eye"],
    "Fashion": ["fashion", "jewelry", "perfume", "cosmetic", "portrait-studio",
                "dress", "suit", "shoes", "glasses", "watch", "bag", "makeup", "hairstyle", "clothes"],
    "Sensitive": ["naked", "nude", "lingerie", "underwear", "swimsuit", "bikini", "beachwear",
                  "boudoir", "sensual", "erotic", "implied-nudity", "topless"],
}

TAG_CANDIDATES: list[str] = [t for tags in TAG_TAXONOMY.values() for t in tags]

TAG_TO_CATEGORY: dict[str, str] = {t: cat for cat, tags in TAG_TAXONOMY.items() for t in tags}


def flatten_taxonomy(taxonomy: dict[str, list[str]] | None) -> list[str]:
    if taxonomy is None:
        return TAG_CANDIDATES
    return [t for tags in taxonomy.values() for t in tags]

File: backend/internal/api/media/taxonomy.go
package media

// tagCategories maps individual CLIP tags to smart-folder category names.
// TODO: Make this configurable (e.g. load from app_config or a dedicated table)
// so users can customize categories without rebuilding.
var tagCategories = map[string]string{
	"person": "People", "people": "People", "portrait": "People", "group": "People",
	"selfie": "People", "baby": "People", "family": "People", "man": "People",
	"woman": "People", "child": "People", "kid": "People", "toddler": "People",
	"elderly": "People", "couple": "People", "crowd": "People", "friends": "People",
	"nature": "Nature", "landscape": "Nature", "mountain": "Nature", "forest": "Nature",
	"tree": "Nature", "flower": "Nature", "plant": "Nature", "garden": "Nature",
	"desert": "Nature", "jungle": "Nature", "canyon": "Nature", "field": "Nature",
	"meadow": "Nature", "hills": "Nature", "cliff": "Nature", "grass": "Nature", "rock": "Nature",
	"ocean": "Water", "beach": "Water", "sea": "Water", "lake": "Water", "river": "Water",
	"water": "Water", "underwater": "Water", "wave": "Water", "coral": "Water",
	"waterfall": "Water", "swimming-pool": "Water", "pond": "Water", "stream": "Water", "creek": "Water",
	"sunset": "Sky & Light", "sunrise": "Sky & Light", "sky": "Sky & Light", "cloud": "Sky & Light",
	"night": "Sky & Light", "moon": "Sky & Light", "star": "Sky & Light", "milky-way": "Sky & Light",
	"aurora": "Sky & Light", "golden-hour": "Sky & Light", "blue-hour": "Sky & Light",
	"twilight": "Sky & Light", "morning": "Sky & Light", "afternoon": "Sky & Light",
	"evening": "Sky & Light", "sunlight": "Sky & Light", "lens-flare": "Sky & Light",
	"bokeh": "Sky & Light", "backlight": "Sky & Light",
	"city": "Urban", "urban": "Urban", "building": "Urban", "architecture": "Urban",
	"bridge": "Urban", "street": "Urban", "road": "Urban", "alley": "Urban",
	"skyline": "Urban", "indoor": "Urban", "room": "Urban", "house": "Urban",
	"home": "Urban", "office": "Urban", "warehouse": "Urban", "studio": "Urban",
	"subway": "Urban", "skyscrapers": "Urban", "monument": "Urban", "cafe": "Urban",
	"restaurant": "Urban", "living-room": "Urban", "bedroom": "Urban",
	"food": "Food", "drink": "Food", "meal": "Food", "kitchen": "Food", "coffee": "Food",
	"dessert": "Food", "fruit": "Food", "bread": "Food", "vegetables": "Food",
	"meat": "Food", "rice": "Food", "pasta": "Food", "pizza": "Food", "burger": "Food",
	"sushi": "Food", "tea": "Food", "beer": "Food", "wine": "Food",
	"animal": "Animals", "pet": "Animals", "dog": "Animals", "cat": "Animals",
	"bird": "Animals", "horse": "Animals", "wildlife": "Animals", "fish": "Animals",
	"butterfly": "Animals", "insect": "Animals", "reptile": "Animals", "deer": "Animals",
	"squirrel": "Animals", "lion": "Animals", "tiger": "Animals", "bear": "Animals", "elephant": "Animals",
	"vehicle": "Vehicles", "car": "Vehicles", "bicycle": "Vehicles", "airplane": "Vehicles",
	"train": "Vehicles", "boat": "Vehicles", "motorcycle": "Vehicles", "truck": "Vehicles",
	"bus": "Vehicles", "scooter": "Vehicles", "ship": "Vehicles", "yacht": "Vehicles", "helicopter": "Vehicles",
	"sport": "Sport", "fitness": "Sport", "running": "Sport", "biking": "Sport",
	"soccer": "Sport", "basketball": "Sport", "swimming": "Sport", "yoga": "Sport",
	"tennis": "Sport", "golf": "Sport", "badminton": "Sport", "baseball": "Sport",
	"gym": "Sport", "climbing": "Sport", "surfing": "Sport", "skiing": "Sport", "skateboarding": "Sport",
}

// CategoryForTag returns the smart-folder category for a given tag.
// Tags not in the mapping return "Other".
func CategoryForTag(tag string) string {
	if cat, ok := tagCategories[tag]; ok {
		return cat
	}
	return "Other"
}
